### ISSUE TITLE
Always changes in tf plan using resource "openstack_keymanager_secret_v1" when payload_content_type = "application/octet-stream"

### ISSUE THREAD

**Issue by lcurtis-datto** [terraform-provider-openstack/terraform-provider-openstack#1492]
Hello all.

We are trying to use Terraform to manage Openstack Barbican secrets for Openstack LBAAS and would like to only upload new secrets when the cert/payload is changed. However, when running tf plan with an unchanged certificate file and payload_content_type is set to "application/octet-stream", a change is forced. Setting an ignore on payload attribute is not an acceptable workaround since we must detect when certificate files are changed. We started seeing this issue when trying to pull certs using  Hashicorp Vault data source, but this issue also occurs even just using a local file.

This issue does not occur when  payload_content_type is set to "text/plain". It looks like maybe the secret is not read back from Openstack properly, so a new payload is generated.

### Terraform Version
Terraform v1.3.9
on linux_amd64
+ provider registry.terraform.io/terraform-provider-openstack/openstack v1.49.0
(Openstack version is Zed)

### Affected Resource(s)
Please list the resources as a list, for example:
- openstack_keymanager_secret_v1

If this issue appears to affect multiple resources, it may be an issue with Terraform's core, so please mention this.

### Terraform Configuration Files
```hcl
terraform {
  required_version = ">= 1.3.0"
  required_providers {
    openstack = {
      source = "terraform-provider-openstack/openstack"
    }
  }
}
resource "openstack_keymanager_secret_v1" "certificate_1" {
  name                 = "certificate"
  payload_content_type = "application/octet-stream"
  payload_content_encoding = "base64"
  payload              = filebase64("my_cert.p12")
  secret_type          = "certificate"
}
```

### Debug Output
https://gist.github.com/lcurtis-datto/737f2a029529da1f3d07810f7dee9577

### Expected Behavior
Running `terraform plan` should have resulted in no changes

### Actual Behavior
Output of "openstack_keymanager_secret_v1.certificate_1 must be replaced"
      + payload                  = (sensitive value) # forces replacement

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform apply`
2. `terraform plan`

### Important Factoids

Looking at the debug output, it seems that the payload_content_type is not being set when trying to read the secret. I can produce similar error using openstack-cli and not specifying payload content-type.

EX: 
`openstack secret get https://<REDACTED>:9311/v1/secrets/11f1812b-8b67-4157-83de-d8a4cc8652ba -p
4xx Client error: Not Found: Secret not found.
Not Found: Secret not found.`

https://gist.github.com/lcurtis-datto/737f2a029529da1f3d07810f7dee9577#file-plan-out-L219

**Comment by lcurtis-datto** [comment]
It looks like this line may be the issue - https://github.com/terraform-provider-openstack/terraform-provider-openstack/blob/cec35ae29769b4de7d84980b1335a2b723ffb15f/openstack/keymanager_secret_v1.go#L100

There needs to be a payload_content_type argument set in order to read the payload if it is not plain-text...not sure if this is a versioning issue with Openstack or not.

**Comment by lcurtis-datto** [comment]
Thanks so much for the quick fix. As a note, I still ran into changes detected in our actual use case since secrets were originally encoded using GNU base64 util which inserts newline characters. This would cause diffs when reading back secret from Openstack. I don't know whether that would need further addressing or not within the provider, but maybe a mention in the docs?

Leaving this here in case anyone runs into this scenario. My workaround is to just use the TF 'replace' function. _Ex:_

`resource "openstack_keymanager_secret_v1" "my_cert" {`
  `name                     = "my_tls_cert"`
  `payload                  = replace(data.vault_generic_secret.my_crt_base64.data["my_crt.p12.base64"],"\n", "")`
  `payload_content_type     = "application/octet-stream"`
  `payload_content_encoding = "base64"`
  `secret_type              = "certificate"`
`}`

Cheers!

**Comment by kayrus** [comment]
@lcurtis-datto right. I haven't though about this. I can keep the previous DiffSuppressFunc.

**Comment by kayrus** [comment]
@lcurtis-datto please check the recent update.

**Comment by lcurtis-datto** [comment]
> @lcurtis-datto right. I haven't though about this. I can keep the previous DiffSuppressFunc.

I am not sure whether to leave in or not. It seems logical to leave in since the resource is indeed computed, as long as end users have a workaround. I will test.

Thanks for the additional comments regarding whitespaces, also.

**Comment by lcurtis-datto** [comment]
Recompiling with [Computed: true](https://github.com/terraform-provider-openstack/terraform-provider-openstack/pull/1505/files#diff-fa51d90cd1d6e411527088513592ace095bbff83ca5dfa7b5791a2d9c5bc0d71L101) still worked in my use case as long as I had the 'replace' function to remove the newlines from the secret. When I removed the function and just used:
`payload  = data.vault_generic_secret.my_crt_base64.data["my_crt.p12.base64"]`

a change was produced.

**Comment by kayrus** [comment]
There is no reason for `Computed: true` and basically the behavior should be the same. Replacing newlines should not affect the state. Can you post more details on your particular case? Which steps should be performed to trigger a change?

**Comment by lcurtis-datto** [comment]
Agreed, and behavior is the same. Here are basic steps in our use case that will trigger a change using an example of example.net x.509 cert and key.

Using [#1505](https://github.com/terraform-provider-openstack/terraform-provider-openstack/pull/1505)

-  Take x.509 key , chained cert and convert to pkcs12 format as required by Openstack LBAAS:
` openssl pkcs12 -export -inkey example.net.key -in example.net.chained.pem -passout pass: -out example.net.p12`
- Encode the p.12 binary to base64:
`base64 < example.net.p12 > example.net.p12.base64`
- Store in Hashicorp Vault
`vault kv put -mount certs example.net  example.net.p12.base64=@example.net.p12.base64`
- Run `terraform apply` to [store Vault secret to Openstack Keymanager](https://gist.github.com/lcurtis-datto/df82e9720ff7fa943fdf4f474c93ecfa) (use code in gist or similar)
- Run `terraform plan`

A Terraform change is flagged unless the  'replace ' function is used, which now makes perfect sense to me due to the way Barbican stores the secret.

The Vault secret contains newline characters due to being encoded by the GNU` base64` util, but Barbican appears to normalize the Base64 data as shown by the test output file. Therefore, the secrets would never match without stripping the newlines from the Vault secret.

Long story short - I think your fix is perfect as-is so long as it passes all of your required tests / earlier Openstack versions. I sincerely appreciate your efforts on this.

**Comment by kayrus** [comment]
> A Terraform change is flagged unless the 'replace ' function is used, which now makes perfect sense to me due to the way Barbican stores the secret.

replace replaces all newline symbols. the Diff hook covers only leading and trailing whitespaces, which may occur when terraform argument is define in a multi line style:
```hcl
resource "openstack_keymanager_secret_v1" "secret_1" {
  algorithm = "aes"
  bit_length = 256
  mode = "cbc"
  name = "mysecret"
  payload = <<EOF
updatedfoobar
EOF
  payload_content_type = "text/plain"
  secret_type = "passphrase"
}
```

**Comment by lcurtis-datto** [comment]
Understood. I mentioned the replace merely as an example.

The PR looks good from my perspective. Looking forward to it.
